File: src/services/productService.js
import envApiService from './envApi.js'

console.log('🛍️ ProductService loading with ENV API (Netlify variables)...')

class ProductService {
  constructor() {
    console.log('🏪 ProductService initialized with envApiService - Reading Netlify env vars!')
  }

  // Get all products with optional filtering - ENV VERSION
  async getProducts(category = null, filters = {}) {
    try {
      console.log('🌟 ENV getProducts called with:', { category, filters })
      
      const params = new URLSearchParams()
      if (category) params.append('category', category)
      Object.keys(filters).forEach(key => {
        if (filters[key]) params.append(key, filters[key])
      })
      
      const queryString = params.toString()
      const endpoint = `/products/${queryString ? `?${queryString}` : ''}`
      
      console.log(`🛍️ ENV Fetching products: ${endpoint}`)
      const response = await envApiService.get(endpoint)
      
      // Handle both array response and object with products array
      const products = Array.isArray(response) ? response : (response.products || [])
      console.log(`✅ ENV Fetched ${products.length} products`)
      
      return { products, pagination: response.pagination }
    } catch (error) {
      console.error('❌ ENV Failed to fetch products:', error)
      return { products: [], pagination: null }
    }
  }

  // Get single product by ID
  async getProduct(id) {
    return envApiService.get(`/products/${id}/`)
  }

  // Create new product (admin only) - Enhanced for working backend
  async createProduct(productData) {
    try {
      console.log('🏗️ Creating product:', productData)
      
      // Ensure proper data formatting for backend
      const formattedData = {
        name: productData.name?.trim(),
        price: parseFloat(productData.price),
        category: productData.category,
        description: productData.description?.trim() || '',
        brand: productData.brand?.trim() || '',
        condition: productData.condition || 'good',
        image: productData.image || '/api/placeholder/400/400'
      }
      
      const response = await envApiService.post('/products', formattedData)
      
      console.log('✅ Product created successfully:', response)
      return response
    } catch (error) {
      console.error('❌ Product creation failed:', error)
      throw error
    }
  }

  // Update existing product (admin only)
  async updateProduct(id, productData) {
    return envApiService.put(`/products/${id}`, productData)
  }

  // Delete product (admin only) - Fixed for CORS preflight
  async deleteProduct(id) {
    try {
      console.log(`🗑️ Deleting product with ID: ${id}`)
      
      // Ensure we're using numeric ID for backend route matching
      const numericId = parseInt(id)
      if (isNaN(numericId)) {
        throw new Error('Invalid product ID - must be numeric')
      }
      
      const response = await envApiService.delete(`/products/${numericId}`)
      
      console.log('✅ Product deleted successfully')
      return response
    } catch (error) {
      console.error('❌ Product deletion failed:', error)
      throw error
    }
  }

  // Search products
  async searchProducts(query, filters = {}) {
    const params = new URLSearchParams({
      q: query || '',
      ...filters
    })
    return envApiService.get(`/products/search?${params}`)
  }

  // Get products by category
  async getProductsByCategory(category) {
    return this.getProducts(category)
  }

  // Get featured products
  async getFeaturedProducts() {
    return envApiService.get('/products/featured')
  }

  // Get new arrivals (use featured products as fallback since backend doesn't have this endpoint)
  async getNewArrivals() {
    return envApiService.get('/products/featured')
  }

  // Get product categories
  async getCategories() {
    return envApiService.get('/products/categories')
  }

  // Get product reviews
  async getProductReviews(productId) {
    return envApiService.get(`/products/${productId}/reviews`)
  }

  // Add product review
  async addProductReview(productId, reviewData) {
    return envApiService.post(`/products/${productId}/reviews`, reviewData)
  }

  // Get related products
  async getRelatedProducts(productId) {
    return envApiService.get(`/products/${productId}/related`)
  }
}

export default new ProductService()

File: backend/app/services/product_service.py
from sqlalchemy.orm import Session
from sqlalchemy import or_, and_
from typing import Optional, List
from app.models import Product, ProductCondition
from app.schemas import ProductCreate, ProductUpdate

class ProductService:
    @staticmethod
    def create_product(db: Session, product_data: ProductCreate) -> Product:
        """Create a new product"""
        db_product = Product(**product_data.dict())
        
        db.add(db_product)
        db.commit()
        db.refresh(db_product)
        return db_product
    
    @staticmethod
    def get_product_by_id(db: Session, product_id: int) -> Optional[Product]:
        """Get product by ID"""
        product = db.query(Product).filter(Product.id == product_id).first()
        if product:
            # Increment view count
            product.views_count += 1
            db.commit()
        return product
    
    @staticmethod
    def get_products(
        db: Session, 
        skip: int = 0, 
        limit: int = 100,
        category: Optional[str] = None,
        search: Optional[str] = None,
        featured: Optional[bool] = None,
        min_price: Optional[float] = None,
        max_price: Optional[float] = None
    ) -> List[Product]:
        """Get products with filtering"""
        query = db.query(Product).filter(Product.is_available == True)
        
        if category:
            query = query.filter(Product.category == category)
            
        if search:
            query = query.filter(
                or_(
                    Product.name.ilike(f"%{search}%"),
                    Product.description.ilike(f"%{search}%"),
                    Product.brand.ilike(f"%{search}%")
                )
            )
            
        if featured is not None:
            query = query.filter(Product.featured == featured)
            
        if min_price is not None:
            query = query.filter(Product.price >= min_price)
            
        if max_price is not None:
            query = query.filter(Product.price <= max_price)
        
        return query.offset(skip).limit(limit).all()
    
    @staticmethod
    def update_product(db: Session, product_id: int, product_data: ProductUpdate) -> Optional[Product]:
        """Update product"""
        product = ProductService.get_product_by_id(db, product_id)
        if not product:
            return None
        
        update_data = product_data.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(product, field, value)
        
        db.commit()
        db.refresh(product)
        return product
    
    @staticmethod
    def delete_product(db: Session, product_id: int) -> bool:
        """Delete product (mark as unavailable)"""
        product = ProductService.get_product_by_id(db, product_id)
        if not product:
            return False
        
        product.is_available = False
        db.commit()
        return True
    
    @staticmethod
    def get_featured_products(db: Session, limit: int = 8) -> List[Product]:
        """Get featured products"""
        return db.query(Product).filter(
            and_(Product.featured == True, Product.is_available == True)
        ).limit(limit).all()
    
    @staticmethod
    def get_categories(db: Session) -> List[str]:
        """Get all product categories"""
        result = db.query(Product.category).filter(Product.is_available == True).distinct().all()
        return [category[0] for category in result]
    
    @staticmethod
    def search_products(db: Session, query: str, limit: int = 20) -> List[Product]:
        """Search products by name, description, or brand"""
        return db.query(Product).filter(
            and_(
                Product.is_available == True,
                or_(
                    Product.name.ilike(f"%{query}%"),
                    Product.description.ilike(f"%{query}%"),
                    Product.brand.ilike(f"%{query}%")
                )
            )
        ).limit(limit).all()
